cirq/sim/density_matrix_simulator.py

```diff
@@ -278,74 +249,43 @@ def _base_iterator(
         )
         if np.may_share_memory(initial_matrix, initial_state):
             initial_matrix = initial_matrix.copy()
-        measured = collections.defaultdict(bool)  # type: Dict[Tuple[cirq.Qid, ...], bool]
+
         if len(circuit) == 0:
             yield DensityMatrixStepResult(initial_matrix, {}, qubit_map, self._dtype)
             return
 
-        state = _StateAndBuffers(
-            len(qid_shape),
-            initial_matrix.reshape(qid_shape * 2) if not is_raw_state else initial_matrix,
+        tensor = initial_matrix.reshape(qid_shape * 2)
+        sim_state = act_on_density_matrix_args.ActOnDensityMatrixArgs(
+            target_tensor=tensor,
+            available_buffer=[np.empty_like(tensor) for _ in range(3)],
+            axes=[],
+            qid_shape=qid_shape,
+            prng=self._prng,
+            log_of_measurement_results={},
         )
 
-        def on_stuck(bad_op: ops.Operation):
-            return TypeError(
-                "Can't simulate operations that don't implement "
-                "SupportsUnitary, SupportsConsistentApplyUnitary, "
-                "SupportsMixture, SupportsChannel or is a measurement: {!r}".format(bad_op)
-            )
-
-        def keep(potential_op: ops.Operation) -> bool:
-            return protocols.has_channel(potential_op, allow_decompose=False) or isinstance(
-                potential_op.gate, ops.MeasurementGate
-            )
-
         noisy_moments = self.noise.noisy_moments(circuit, sorted(circuit.all_qubits()))
-
+        measured = collections.defaultdict(bool)  # type: Dict[Tuple[cirq.Qid, ...], bool]
         for moment in noisy_moments:
-            measurements = collections.defaultdict(list)  # type: Dict[str, List[int]]
-
-            channel_ops_and_measurements = protocols.decompose(
```

Comment:
Why did you remove this decomposition step and `on_stuck` and `keep`? I am surprised that this wasn't covered by tests -- but it seems that we lost the useful error message above.

E.g.: 

```
import cirq

class TestGate(cirq.Gate):
    def _num_qubits_(self) -> int:
        return 1

a, b, c = cirq.LineQubit.range(3)

circuit = cirq.Circuit(cirq.X(a), TestGate()(b), cirq.measure(c))
print(circuit)

cirq.DensityMatrixSimulator().run(circuit)
```

would fail in master with:

```
TypeError: Can't simulate operations that don't implement SupportsUnitary, SupportsConsistentApplyUnitary, SupportsMixture, SupportsChannel or is a measurement: <__main__.TestGate object at 0x7fea86010610>.on(cirq.LineQubit(1))
```

However it fails with a significantly more obscure message on this branch: 

```
TypeError: Failed to act action on state argument.
Tried both action._act_on_ and args._act_on_fallback_.

State argument type: <class 'cirq.sim.act_on_density_matrix_args.ActOnDensityMatrixArgs'>
Action type: <class 'cirq.ops.gate_operation.GateOperation'>
Action repr: <__main__.TestGate object at 0x7fc15560ce50>.on(cirq.LineQubit(1))
```

Reply:
I followed the pattern established in the original act_on protocol PR, where those were removed for the sparse simulator https://github.com/quantumlib/Cirq/pull/3019/files (on_stuck was removed there and still does not exist in sparse simulator code today). If we want to restore the nice error message I'll look into how to do it for the other simulators too.

Reply:
Looked further, this is going to take some work. The decompose method that is called in the established act_on classes returns the qubits (which we require) but does not handle a keep/on_stuck. The decompose methods that accept keep/onstuck do not return the qubits. Will figure out if this can be fixed.

Reply:
Actually I think this goes back to the "iterate through the list of strategies". The only error message we can provide is that `_act_on_` failed. Because if we return the fact that decompose failed, well the user may not have expected to even hit the decompose strategy and the decomposition error may be irrelevant. Will have to think about this some more.

Reply:
We don't actually need `keep` because the act_on implementation is to do this recursively: act_on will decompose a level, then act_on that level, which will decompose further if necessary. This is different from how things were implemented prior to act_on, which required a perfect decomposition before simulating each op. So, returning the error message if the `_act_on_fallback_` doesn't succeed fixes the problem.